ignite/contrib/handlers/throughput_benchmark_wrapper.py

```diff
@@ -0,0 +1,244 @@
+    ):
+        """
+        Attaches throughput_benchmark to a given engine.
+        As for a benchmark not the complete dataset is required, and could consume too much memory,
+        only the first `max_batches` are used for the benchmark.
+        `input_transform` is used to retrieve the corresponding input for the forward function
+        of the model from the batch.
+
+        Args:
+            engine (Engine): throughput_benchmark is attached to this engine. Please, keep in mind that all
+                attached handlers will be executed.
+            max_batches (int, optional): Number of batches to use for the benchmark. After sampling
+                this many batches, the engine will terminate early if there are remaining iterations.
+                To sample complete dataset use: `len(dataloader)`.
+                Defaults to 10.
+            input_transform (Callable, optional): function that transform the batch from the dataloader
+                to the input required for the model input. Defaults to lambda input_batch:input_batch[0].
+
+        Note:
+            The batches are processed from the `engine.state`. If the input data is different from the
+            data from the dataloader, the wrong data will be benchmarked. Either provide a dataloader
+            with all of the transformations, or include the transformations in `input_transform`.
+
+        Returns:
+            Engine: engine_with_throughput_benchmark: engine used for testing the throughput
+
+        Example:
+
+            .. code-block:: python
+
+                with throughput_bechmark.attach(engine, max_batches=5) as engine_with_throughput_benchmark:
+                    engine_with_throughput_benchmark.run(dataloader)
+
+        """
+
+        def stop_after_max_batches(engine: Engine, event: Events):
+            # Events start with 1
+            if event <= max_batches:
+                if event == max_batches:
+                    engine.terminate()
+                return True
+            return False
+
+        if not isinstance(max_batches, int):
+            raise TypeError("Max batches has to be int. Given: {}".format(type(max_batches)))
+        if not callable(input_transform):
+            raise TypeError("Input transform has to be Callable. Given: {}".format(type(input_transform)))
+        if max_batches < 1:
+            raise ValueError("Max batches to sample from has to be larger than 0. Given: {}".format(max_batches))
+
+        if not engine.has_event_handler(self._run):
+            engine.add_event_handler(
+                Events.ITERATION_STARTED(event_filter=stop_after_max_batches), self._batch_logger, input_transform,
+            )
+            engine.add_event_handler(
+                Events.COMPLETED, self._run,
+            )
+
+        yield engine
+        self._detach(engine)
+
+    @property
+    def stats(self) -> ExecutionStats:
```

Comment:
Maybe we can promote `self._stats` to a public attribute without property wrapping ?
Thus we can avoid redefining `ExecutionStats` and this typing issue.

Reply:
In that case, I personally would still favor writing it like:
`self.stats: Optional[ExecutionStats]`
I am a big fan of type hinting and abusing auto-complete as much as possible ;)

But I agree that the current approach is a sup-bar solution.
In the end, either the class has to be tried to be imported, a stub has to be created or no typing support at all.

But if you like it more that way, I will rewrite the code and update the code to link to the specific class.

Reply:
> I am a big fan of type hinting and abusing auto-complete as much as possible ;)

@kai-tub that's cool ! So, can we find a solution to avoid the stub `ExecutionStats`, keep our `try except` in `__init__` to import pytorch's ThroughputBenchmark  and define the type of self.stats as you propose ?

Reply:
See my new commit. It looks better, but I won't guarantee that all auto-complete tools will work with this annotation style. ;)

Reply:
Totally missed that Python 3.5 is still supported. Yeah, then the variable annotation won't work, so we are back where we started.
Instead of writing a typing stub we could try to import this specific module and catch the import error if present. Since it is never used, it shouldn't cause any problems but it would be a confusing line to read

Reply:
Yeah, 3.5 is a pain for typing.

> Instead of writing a typing stub we could try to import this specific module and catch the import error if present. Since it is never used, it shouldn't cause any problems but it would be a confusing line to read

Could you please give an idea of that with a code ? Otherwise, I'm OK with
```
self.execution_stats = None
```

Reply:
> Could you please give an idea of that with a code ?

Ups, yes

I would go back to using the property
```
    @property
    def stats(self) -> ExecutionStats:
```
Because then ExecutionStats could be annotated in Python 3.5 and instead of creating a stub trying to import the original class.
```
try:
    from torch.utils.throughput_benchmark import ExecutionStats  # for typing
except ImportError:
    pass  # Only needed for type annotation
```

Reply:
Can't we use something like that 
```python
self.execution_stats = None  # type: ExecutionStats
```
for 3.5, 3.6, etc ?

Reply:
Yes, we could use something like that.
What I was trying to go for was a way for the user to load the library without any exceptions but raising a runtime error when the class is used similar to `ml_flow_logger.py` and other handlers.

Just out of curiosity, could you explain why you prefer an accessible attribute over a property wrapped one? Then we could use the *normal* function annotation again. :)

Reply:
> Just out of curiosity, could you explain why you prefer an accessible attribute over a property wrapped one? Then we could use the normal function annotation again. :)

Do I understand correctly your idea with property and typing ?
```python

from typing import Callable, Optional
import contextlib

import torch

from ignite.engine import Events, Engine

# IMO this is as you said looks weird
try:
    from torch.utils.throughput_benchmark import ExecutionStats  # for typing
except ImportError:
    pass  # Only needed for type annotation

class ThroughputBenchmarkWrapper:

     @property
         def stats(self) -> ExecutionStats:
```

Reply:
Sorry for not committing code, I am currently changing everything to use the engine design.
What I am currently thinking of (combining both approaches) is something like this:
```python
from typing import Callable, Optional
import contextlib

import torch

from ignite.engine import Events, Engine

try:
    from torch.utils.throughput_benchmark import ThroughputBenchmark
    from torch.utils.throughput_benchmark import ExecutionStats  # for typing
except ImportError:
    raise RuntimeError(...)

class ThroughputBenchmarkWrapper:

     @property
         def stats(self) -> ExecutionStats:
```
The reason why I didn't write it like that to start out with, was that I tried the user to be able to import the code without raising an exception. To be able to provide typing support, not raising the exception at import time and not using a stub I've proposed the weird try and catch that you just wrote. :)

Reply:
> Sorry for not committing code, I am currently changing everything to use the engine design.

No problems, thanks a lot for working on that !

The issue with 
```python
from typing import Callable, Optional
import contextlib

import torch

from ignite.engine import Events, Engine

try:
    from torch.utils.throughput_benchmark import ThroughputBenchmark
    from torch.utils.throughput_benchmark import ExecutionStats  # for typing
except ImportError:
    raise RuntimeError(...)


class ThroughputBenchmarkWrapper:
    ...
```
as you say, users with pytorch 1.1.0 wont be able to use `ignite.contrib` module at all as it imports `ThroughputBenchmarkWrapper`. So, this solution is not possible.
That's why property seems like to be unfeasible with all our limitations...

Anyway, currently, contrib module is not using typing at all and even core module wont pass `mypy` tests. IMO, the solution is either use typing in the comment or remove the typing (which is :( ) or something else if possible to satisfy all our limitations...

Reply:
Ahhh. I totally missed that because of `__init__.py` the exception will be raised as soon as ignite.contrib is loaded... Ok, then I agree and will use the typing in the comment. :)

I've pushed the redesigned engine version, without any tests or documentation, just so it is easier for us to discuss the code. This version can be called exactly as you previously proposed, but the code feels... Hacky. I've inherited from `Engine` to be able to make changes to how the run function is executed and to be able to initialize the other parameters.

But I am unsure on how to rewrite the code.

Reply:
Yes, I understand your feeling. I feel the same everytime while inheriting directly from Engine and while setup some internal handlers. Maybe we do not need to create any class like `BenchmarkEngine(Engine)` and just have `create_throughput_benchmark` that returns specifically configured instance of Engine ?

Reply:
Hmm.
So this is an alternative idea:
I've used a singleton wrapper, to only create the benchmark once, instead of having
to wrap the complete function again to keep the state.

```python
class SingletonDecorator:
    def __init__(self, class_):
        self.class_ = class_
        self.instance = None

    def __call__(self, *args, **kwargs):
        if self.instance is None:
            self.instance = self.class_(*args, **kwargs)
        return self.instance


def create_throughput_benchmark(
    model: torch.nn.Module,
    device: Optional[Union[str, torch.device]] = None,
    non_blocking: bool = False,
    prepare_input: Callable = _prepare_input,
    max_batches: int = 10,
    num_calling_threads: int = 1,
    num_warmup_iters: int = 10,
    num_iters: int = 100,
) -> Engine:
    try:
        from torch.utils.throughput_benchmark import ThroughputBenchmark
    except ImportError:
        raise RuntimeError("This method requires at least pytorch version 1.2.0")

    ThroughputBenchmark = SingletonDecorator(ThroughputBenchmark)

    def _run(engine: Engine, batch: Sequence[torch.Tensor]) -> None:
        benchmark = ThroughputBenchmark(model)
        # In this loop nothing happens besides the loading of the input data
        model.eval()  # Should the user expect the automatic setting to eval?
        input_data = _prepare_input(batch, device=device, non_blocking=non_blocking)
        benchmark.add_input(input_data)
        if engine.state.iteration == max_batches:
            engine.terminate()
            engine.results = benchmark.benchmark(
                num_calling_threads=num_calling_threads, num_warmup_iters=num_warmup_iters, num_iters=num_iters
            )

    engine = Engine(_run)

    # see PR https://github.com/pytorch/ignite/pull/835
    if device is not None:

        @engine.on(Events.STARTED)
        def move_device(engine):
            model.to(device)

    return engine
```
Now:
- The user has to specify the `max_batches, num_calling_threads` etc. when creating the engine and not when calling the `run` function, as this is now a *vanilla* run function.
- Kept the internal handlers to a minimum and moved the logic to `_run` function which will be called from the engine. 
- **Problem**: I am not sure how the user should access the results... What would be the easiest way? I've played around with the `Engine` a bit, but I cannot change the internal state representation. I've also thought about somehow integrating it into a Metric, so it could be saved there, but that seems to be overkill.

I am sorry, but I need some help to understand how the result could be integrated into the current engine.

PS: Do you like it better, when I paste these "pseudo-examples" here or would you prefer them on my branch? I am hesitating because I don't want to unnecessarily spam the CI pipeline for some random thoughts.